### PR TITLE
Read paths from streams and config from namelist

### PR DESCRIPTION
**Pull request by xylar**
This merge reads parameters from the namelist (instead of from NetCDF attributes) and
paths for model output to be analyzed from the streams file instead of hard coding them.

This merge also adds the ability to read a list of files that match a template from the streams
file (`StreamsFile.readpath`).  Optionally, the list of files can be restricted to lie between a
startDate and endDate.

Two new options, `analysis_start_date` and `analysis_end_date` have been added to the config
file (and given default values of `0001-01-01_00:00:00` and `9999-12-31_23:59:59`, respectively).
These values can be modified to allow analysis of a subset of output files (e.g. for debugging).

### CODE REVIEW

**Comment by xylar** [comment]
This merge branches off of #37, so should only be merged after that PR has been fulfilled.

**Comment by xylar** [comment]
@pwolfram and @milenaveneziani 
For now, only the ocean heat content time-series analysis function has been updated.
This is intended to serve as an example from which the other analysis functions can be
updated in a similar fashion.

Feel free to update other analysis functions and push the changes to this branch.

**Comment by xylar** [comment]
I tested this on Edison using the config file at:
```
/project/projectdirs/acme/www/xylar/config.xylar on Edison
```
The relevant contents are:
```
[case]
# ACME case configuration
casename = 20160923.alpha8.A_WCYCL1850.ne30_oEC.edison
native_res = ne30
short_term_archive = 0
ref_casename_v0 = B1850C5_ne30_v0.4

[input]
# names of namelist and streams files
namelist_filename = mpas-o_in
streams_filename = streams.ocean

[paths]
# paths to simulation and observational datasets
archive_dir = /global/cscratch1/sd/golaz/ACME_simulations
archive_dir_ocn = /global/cscratch1/sd/golaz/ACME_simulations/20160923.alpha8.A_WCYCL1850.ne30_oEC.edison/run
scratch_dir = /scratch1/scratchdirs/xylar/acme/testing_xylar/scratch
plots_dir = /scratch1/scratchdirs/xylar/acme/testing_xylar/plots
log_dir = /scratch1/scratchdirs/xylar/acme/testing_xylar/logs
obs_ocndir = /global/project/projectdirs/acme/observations/Ocean
obs_sstdir = /global/project/projectdirs/acme/observations/Ocean/SST
obs_seaicedir = /global/project/projectdirs/acme/observations/SeaIce
ref_archive_v0_ocndir = /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ocn/postprocessing
ref_archive_v0_seaicedir = /global/project/projectdirs/acme/ACMEv0_lowres/B1850C5_ne30_v0.4/ice/postprocessing

[data]
# paths to mesh and mapping files
mpas_meshfile = /global/project/projectdirs/acme/milena/MPAS-grids/ocn/gridfile.oEC60to30.nc
mpas_remapfile = /global/project/projectdirs/acme/mapping/maps/map_oEC60to30_TO_0.5x0.5degree_blin.160412.nc
pop_remapfile = /global/project/projectdirs/acme/mapping/maps/map_gx1v6_TO_0.5x0.5degree_blin.160413.nc
# path to climotology dataset
mpas_climodir = /global/project/projectdirs/acme/xylar/20160805v0atm.A_WCYCL1850_v0atm.ne30_oEC.edison.alpha7_00.test.pp

[seaIceData]
# paths to sea-ice observational datasets
obs_iceareaNH = /global/project/projectdirs/acme/observations/SeaIce/IceArea_timeseries/iceAreaNH_climo.nc
obs_iceareaSH = /global/project/projectdirs/acme/observations/SeaIce/IceArea_timeseries/iceAreaSH_climo.nc
obs_icevolNH = /global/project/projectdirs/acme/observations/SeaIce/PIOMAS/PIOMASvolume_monthly_climo.nc
obs_icevolSH = none

[time]
# the first year over which to average climotologies
climo_yr1 = 6
# the last year over which to average climotologies
climo_yr2 = 10
# the offset year to be added to simulation years
yr_offset = 1849

[ohc_timeseries]
generate = 1

[sst_timeseries]
generate = 0

[nino34_timeseries]
generate = 0

[mht_timeseries]
generate = 0

[moc_timeseries]
generate = 0

[sst_modelvsobs]
generate = 0

[seaice_timeseries]
generate = 0

[seaice_modelvsobs]
generate = 0
```

**Comment by xylar** [comment]
This PR is intended to partially address #20 and #32.

**Comment by xylar** [comment]
I added testing of the new `readpath` method in `StreamsFile`.  I can confirm that the test passed for me.  It would be good to add a more rigorous test involving fmatch, which I will work on now.

**Comment by xylar** [comment]
@pwolfram, this should now be ready for you to take a look.

**Comment by xylar** [comment]
@milenaveneziani, could you let me know if you'd rather merge this as is or make similar modifications to the other analysis functions?  I'm fine either way.  I could also probably find time to do the changes myself tomorrow but figured it might be a good test of the code if you can make the necessary modifications.

**Comment by xylar** [comment]
By the way, I don't know why GitHub is showing 3 commits for this branch.  I just did a test merge and it didn't have a problem, so I wouldn't worry about it.  It just makes the reviewing confusing.  I'll rebase if it makes your lives easier.

**Comment by milenaveneziani** [comment]
@xylar: I would like to add other modifications as well. I should be able to work on this tomorrow.
As far as diagnosing the alpha8 run that Jon is running by tomorrow, I will work locally out of @vanroekel branch for #29, and hardwire the  filename changes that we had in alpha8.
What do you think?

**Comment by xylar** [comment]
> @xylar: I would like to add other modifications as well. I should be able to work on this tomorrow.

Great.  I'll see see if I have time to get to some of it tomorrow before you get started.  I'll make sure to push whatever I have so you can pick up where I leave off.

> As far as diagnosing the alpha8 run that Jon is running by tomorrow, I will work locally out of @vanroekel branch for #29, and hardwire the filename changes that we had in alpha8.
What do you think?

That sounds fine as long as those changes don't get pushed up to develop.

**Comment by xylar** [comment]
@milenaveneziani, I'm currently working on adding similar functionality to the two SST functions.

**Comment by vanroekel** [comment]
@xylar, @pwolfram, and @milenaveneziani given the improvements to sst_modelvsobs in this pr, should #29 be modified to be consistent with this PR?  The merge of #29 will remove sst_modelvsobs.py.  @xylar, if you'd like I would be happy to implement your changes from this PR into the new and more general ocean_modelvsobs.py.

**Comment by milenaveneziani** [comment]
@vanroekel (and you and others should correct if I am wrong): it seems to me that we should first merge #29 as is, and then work on the modelvsobs general routine within this PR. I like the idea of having all modifications that have to do with reading namelists/streams in this PR.

**Comment by xylar** [comment]
@milenaveneziani, CSCRATCH on edison just went down so I didn't finish testing my changes.  With IC still down, I'm not sure what I can do about this.  Any suggestions?

**Comment by xylar** [comment]
@vanroekel and @milenaveneziani, yes, let's merge #29 first and I (or one of you) will update this PR accordingly.  That shouldn't be difficult, though it was silly of me to make conflicting modifications.

**Comment by xylar** [comment]
I'm thinking of rebasing this branch onto #29 so there's less work once #29 gets merged.  Any objections?

**Comment by milenaveneziani** [comment]
Sounds good @xylar. There isn't much we can do about testing for now. We could move to OLCF, where I have data from the RRS30to10 runs, but we would need to test/change run_RHEA first. Wait, no, rhea is also down today...

**Comment by xylar** [comment]
...and I don't have an account on Rhea anyway (as far as I know) so that would slow us down, too.

**Comment by xylar** [comment]
I rebased onto #29, which may mean I need to rebase again after #29 gets merged.

**Comment by pwolfram** [comment]
@xylar, I suspect you will have to do the rebase now that #29 is merged.  Thanks for your patience while I've been traveling.

**Comment by xylar** [comment]
@milenaveneziani and/or @vanroekel, could one of you test to make sure this works as intended?  Keep in mind that it doesn't yet have the option to set start and end dates for the analysis (see #20).

**Comment by xylar** [comment]
@vanroekel, it looks like you don't have the namelist or streams files in the data you restored into your scratch folder (https://github.com/MPAS-Dev/MPAS-Analysis/pull/29#issuecomment-256716588), meaning it won't work with this branch.  If you get a chance (and you have those files) could you restore them from HPSS as well?

**Comment by vanroekel** [comment]
@xylar, I took another look.  Unfortunately I don't see the streams or namelist files in Chris Golaz's directory on hpss.  It looks like the nc files were only saved...

**Comment by xylar** [comment]
@vanroekel, I noticed that, too, for some of the runs.  We'll have to make sure Chris knows that we need the streams and namelist files to be preserved from now on.

**Comment by xylar** [comment]
@milenaveneziani, @vanroekel and @pwolfram, I went ahead and implemented the ability to set a start and end date for the analysis, and the ability to limit the list of files found through templates in streams files to this range.

I haven't had a chance to test these changes because of CSCRATCH being down on Edison.  I'll test as soon as it comes back up.  In the meantime, I'm putting the `don't merge` flag back in place.

@pwolfram, I've added some more tests and verified that they pass pytest.  Let me know if you have any concerns or feedback on the tests.  It's actually a rather fun way of working!  You can be pretty sure what you did worked right without having to run a complicated setup!

**Comment by xylar** [comment]
I believe this branch should now fully address #36

**Comment by pwolfram** [comment]
@xylar, is this PR ready for review?  It still has the `don't merge` flag.  I'm glad that you are finding pytest useful.  I'm thinking we are getting to the point where a continuous-integration solution could also be helpful.  I'll make an issue that we should add automatic CI.

**Comment by xylar** [comment]
@pwolfram, I haven't had a chance to test this since CSCRATCH came back up.  You can test it but I'd like to test it myself before it gets merged and I think @milenaveneziani should probably also test it.

**Comment by milenaveneziani** [comment]
I am getting to it now.

**Comment by xylar** [comment]
@milenaveneziani, I need to do some debugging, so no point in you testing just yet, as it turns out.  Sorry about that.

Do feel free to take a look at the code, though, and make sure it makes sense to you.

File: config.analysis [hunk 1]
```diff
@@ -41,6 +48,9 @@ climo_yr1 = 6
 climo_yr2 = 10
 # the offset year to be added to simulation years
 yr_offset = 1849
+# start and end date for the analysis
+analysis_start_date = 0001-01-01_00:00:00
+analysis_end_date = 9999-01-01_00:00:00
```
Comment (xylar) [inline config.analysis]: Thanks, @milenaveneziani.  I can definitely make these changes.  Before I do, I just want to make sure you don't foresee the need to specify start and end times that are not at year boundaries.  I guess I see this restriction as appropriate for ACME but rather limiting for the MPAS-Analysis when it's used more generally.

I guess what I'll do is go along with this plan for this PR, but it's something I think we should rethink in the future.

Comment (milenaveneziani) [inline config.analysis]: Here is perhaps a better option: 1) leave timeseries_start_date and timeseries_end_date in, 2) use *only* the years to select the files, and 3) use the full dates to let xarray to bracket the exact time period. This way we avoid counting too much on filename time stamps, but also have the option to choose specific time periods.

Comment (xylar) [inline config.analysis]: @milenaveneziani, I think I have a compromise solution.  If fields `climo_yr1`, `climo_yr2`, `timeseries_yr1` and `timeseries_yr2` exist, I will use them set default values of fields `climo_start_date`, `climo_end_date`, `timeseries_start_date` and `timeseries_end_date` in `run_analysis.py`.  But the longer-term goal would be to support general values for `*start_date` and `*end_date` that aren't just year boundaries.

**Comment by xylar** [comment]
@vanroekel, the MLD observations files don't appear to be on Edison. Could you copy them there so I can test that part of the code?  Thanks!

**Comment by xylar** [comment]
@milenaveneziani, it appears you're the only one with write access to the folder
```
/global/project/projectdirs/acme/observations/Ocean/
```
Could you copy the MLD observations there, instead, since @vanroekel won't be able to?

**Comment by milenaveneziani** [comment]
sure. And I will change permissions also.

**Comment by vanroekel** [comment]
@milenaveneziani you can grab the MLD observations from my home directory /global/homes/l/lvroekel/MLD, permissions should be okay

**Comment by milenaveneziani** [comment]
All MLD transferred to:
/global/homes/m/milena/proj_acme/observations/Ocean/MLD/
and I changed the permissions too.

**Comment by xylar** [comment]
@milenaveneziani, it looks like the permissions still aren't good on all the files.  can you recursively change group to acme and recursively given full read/write permission?

File: config.analysis (hunk 1 above)
Comment (xylar) [inline config.analysis]: @milenaveneziani, I didn't notice your comment from earlier.
> Here is perhaps a better option: 1) leave timeseries_start_date and timeseries_end_date in, 2) use only the year values to select the files, and 3) use the full dates to let xarray to bracket the exact time period. This way we avoid counting too much on filename time stamps, but also have the option to choose specific time periods.

I see, you're worried that the dates in file names aren't a sure way to determine which dates are *actually* in a given file.  That's a tricky one.

**Comment by milenaveneziani** [comment]
ok, the group had been fixed. sorry about that.

File: config.analysis (hunk 1 above)
Comment (xylar) [inline config.analysis]: Got it.  So basically either way we need to make sure we have some buffer where we search for files outside the specified range.  I guess the buffer has to come from the file writing interval.  I'll see what can be done, but it should be possible to try reading one file too many in both directions to be sure we have all the correct dates, then `mpas_xarray` can sort out the details.

Comment (xylar) [inline config.analysis]: Down the road, one option would be a quick pre-processing step (presumably one would be needed for each stream we use) that would go through all files from that stream, read all xtime values, and cache a list of only those files with xtime within the desired range.  Too much for this PR, though.

**Comment by xylar** [comment]
@milenaveneziani, I'm trying to run the analysis tools in `develop` as a baseline for making sure this PR is behaving correctly. I'm using the alpha7 output you recommended yesterday.  Here's my config file:
https://gist.github.com/xylar/228bdcd9ef4d19fd823ec33525e1ec42
(Ignore any options in there that are only valid for this PR -- they do no harm when running with `develop`.)  Everything works (doesn't crash) up to sea ice climatology.  Then, I get the following error:
```
Plotting 2-d maps of sea-ice concentration and thickness climatologies...

Climatology directory does not exist. Create it...


Regridded directory does not exist. Create it...

/global/cscratch1/sd/golaz/ACME_simulations/20160805.A_WCYCL1850.ne30_oEC.edison.alpha7_00/run
/scratch1/scratchdirs/xylar/acme/baseline/climo
  Load sea-ice data...
/global/cscratch1/sd/golaz/ACME_simulations/20160805.A_WCYCL1850.ne30_oEC.edison.alpha7_00/run
  Compute seasonal climatologies...
  Regrid fields to regular grid...
basename: missing operand
Try `basename --help' for more information.
Started processing at Wed Nov  2 03:16:15 PDT 2016.
Running remap script  from directory /global/u2/x/xylar/mpas_work/analysis/mpas_analysis
NCO version "4.6.2-alpha01" from directory /global/u1/z/zender/bin_edison
Input files in or relative to directory /global/homes/x/xylar/mpas_work/analysis/mpas_analysis
Intermediate/temporary files written to directory /tmp
Output files to directory /scratch1/scratchdirs/xylar/acme/baseline/climo/mpas_regridded
Input data obey "mpas"-ordering and will be permuted with "ncpdq -a Time,nVertLevels,maxEdges,MaxEdges2,nCategories,ONE,nEdges,nCells"
Input assumed to be MPAS-O/I data: will apply renormalized regridding (--rnr=0.0)
Input assumed to be barenaked of attributes: will annotate NC_DOUBLE variables with _FillValue prior to regridding
Source and destination grids will both be read from supplied map-file
Input #00: /scratch1/scratchdirs/xylar/acme/baseline/climo/mpas-cice_climo.years0006-0010.on.nc
PDQ(in)  : /tmp/ncremap_tmp_att.nc.pid38547
PDQ(out) : /tmp/ncremap_tmp_pdq.nc.pid38547
/usr/common/software/nco/4.6.1/intel/bin/ncremap: line 1058: 38753 Segmentation fault      ncpdq -O --no_tmp_fl --gaa remap_script= --gaa remap_hostname=edison08 --gaa remap_version="4.6.2-alpha01" --hdr_pad=1000 -a Time,nVertLevels,maxEdges,MaxEdges2,nCategories,ONE,nEdges,nCells /tmp/ncremap_tmp_att.nc.pid38547 /tmp/ncremap_tmp_pdq.nc.pid38547
: ERROR Failed to generate pdq-file. Debug this:
ncpdq -O --no_tmp_fl --gaa remap_script= --gaa remap_hostname=edison08 --gaa remap_version="4.6.2-alpha01" --hdr_pad=1000  -a Time,nVertLevels,maxEdges,MaxEdges2,nCategories,ONE,nEdges,nCells /tmp/ncremap_tmp_att.nc.pid38547 /tmp/ncremap_tmp_pdq.nc.pid38547
Error with call  ncremap -P mpas -i /scratch1/scratchdirs/xylar/acme/baseline/climo/mpas-cice_climo.years0006-0010.on.nc -m /global/project/projectdirs/acme/mapping/maps/map_oEC60to30_TO_0.5x0.5degree_blin.160412.nc -O /scratch1/scratchdirs/xylar/acme/baseline/climo/mpas_regridded
Command '['ncremap', '-P', 'mpas', '-i', '/scratch1/scratchdirs/xylar/acme/baseline/climo/mpas-cice_climo.years0006-0010.on.nc', '-m', '/global/project/projectdirs/acme/mapping/maps/map_oEC60to30_TO_0.5x0.5degree_blin.160412.nc', '-O', '/scratch1/scratchdirs/xylar/acme/baseline/climo/mpas_regridded']' returned non-zero exit status 1
Traceback (most recent call last):
  File "run_analysis.py", line 133, in <module>
    analysis(config)
  File "run_analysis.py", line 110, in analysis
    seaice_modelvsobs(config)
  File "/global/u2/x/xylar/mpas_work/analysis/mpas_analysis/mpas_analysis/sea_ice/modelvsobs.py", line 139, in seaice_modelvsobs
    raise e
subprocess.CalledProcessError: Command '['ncremap', '-P', 'mpas', '-i', '/scratch1/scratchdirs/xylar/acme/baseline/climo/mpas-cice_climo.years0006-0010.on.nc', '-m', '/global/project/projectdirs/acme/mapping/maps/map_oEC60to30_TO_0.5x0.5degree_blin.160412.nc', '-O', '/scratch1/scratchdirs/xylar/acme/baseline/climo/mpas_regridded']' returned non-zero exit status 1
```
I'll look into it but would appreciate any help you might be able to give me.  Do I have some path wrong?

**Comment by xylar** [comment]
@milenaveneziani, I found that the problem seems to be with `nco/4.6.1` (the default on Edison).  If I load `nco/4.6.0`, the same command runs successfully.  If someone else can reproduce this problem, we should probably let Charlie Zender know about it, as this will affect ACME analysis on Edison generally, not just us.

File: config.analysis (hunk 1 above)
Comment (pwolfram) [inline config.analysis]: @xylar, I may be missing something here but what is the risk of opening too many files via xarray (which does lazy evaluation) and then segmenting in time following that step?   The reason I ask is because the above discussion may result in unnecessary code unless the loading step is demonstrated to be the bottleneck.  My understanding is that essentially xarray takes care of this heavy lifting behind the scenes.

**Comment by xylar** [comment]
Just a quick note to say that this branch is currently a bit of a mess.  I'm working on it but please ignore for now...

File: mpas_analysis/ocean/ocean_modelvsobs.py [hunk 2]
```diff
@@ -41,7 +59,7 @@ def ocn_modelvsobs(config, field):
 
     #Seems like the following line should be a config.get option and not
     #read every time series file when only a subset is taken
-    infiles = "".join([indir,"/am.mpas-o.timeSeriesStats.????-*.nc"])
+    #infiles = "".join([indir,"/am.mpas-o.timeSeriesStats.????-*.nc"])
```
Comment (xylar) [inline mpas_analysis/ocean/ocean_modelvsobs.py]: @pwolfram, I agree.  I was keeping them to help me debug for the time being.  Please don't worry about reviewing the code right now.  As I just posted, it's very much a work in progress based on @milenaveneziani's and my recent discussion of what features are needed.

File: mpas_analysis/ocean/ohc_timeseries.py [hunk 3]
```diff
@@ -44,7 +81,6 @@ def ohc_timeseries(config):
 
     # Load data
     print "  Load ocean data..."
-    infiles = '%s/am.mpas-o.timeSeriesStats.????-??*nc'%indir
 
     # Load data:
     ds = xr.open_mfdataset(infiles, preprocess=lambda x: preprocess_mpas(x, yearoffset=yr_offset,
```
Comment (pwolfram) [inline mpas_analysis/ocean/ohc_timeseries.py]: I think we can clean up the above lines to 

```python
     # Load data
     print "  Load ocean data..."
     ds = xr.open_mfdataset(infiles, preprocess=lambda x: preprocess_mpas(x, yearoffset=yr_offset,
```
but this is a nit.

File: mpas_analysis/ocean/sst_timeseries.py [hunk 4]
```diff
@@ -1,23 +1,47 @@
...
+    streams = StreamsFile(streams_filename, streamsdir=indir)
+
+    # get a list of timeSeriesStats output files from the streams file,
+    # reading only those that are between the start and end dates
+    startDate = config.get('time', 'timeseries_start_date')
+    endDate = config.get('time', 'timeseries_end_date')
+    infiles = streams.readpath('timeSeriesStatsOutput', 'filename_template',
+                               startDate=startDate, endDate=endDate)
+    print 'Reading files {} through {}'.format(infiles[0],infiles[-1])
+
```
Comment (pwolfram) [inline mpas_analysis/ocean/sst_timeseries.py]: @xylar, is it possible to generalize the above code across multiple files?  This would produce a cleaner PR that would be easier to maintain for the future.  I'm happy to help with this if we all agree it is useful.

File: mpas_analysis/sea_ice/modelvsobs.py [hunk 5]
```diff
@@ -90,7 +117,7 @@ def seaice_modelvsobs(config):
     print "  Load sea-ice data..."
     print indir
 
-    infiles = "".join([indir,"/am.mpas-cice.timeSeriesStatsMonthly.????-*.nc"])
+    #infiles = "".join([indir,"/am.mpas-cice.timeSeriesStatsMonthly.????-*.nc"])
     #infiles = "".join([indir,"/am.mpas-cice.timeSeriesStatsMonthly.001[5-6]-*.nc"])
```
Comment (pwolfram) [inline mpas_analysis/sea_ice/modelvsobs.py]: Don't need lines of code above.

File: mpas_analysis/sea_ice/timeseries.py [hunk 6]
```diff
@@ -48,7 +75,7 @@ def seaice_timeseries(config):
     dsmesh = xr.open_dataset(meshfile)
 
     # Load data
-    infiles = "".join([indir,'/am.mpas-cice.timeSeriesStatsMonthly.????-??-??.nc'])
+    #infiles = "".join([indir,'/am.mpas-cice.timeSeriesStatsMonthly.????-??-??.nc'])
```
Comment (pwolfram) [inline mpas_analysis/sea_ice/timeseries.py]: Probably can remove.

File: mpas_analysis/test/test_namelist_streams_interface.py [hunk 7]
```diff
@@ -47,4 +49,54 @@ def test_read_streamsfile(self):
...
+                                    endDate='0001-12-31')
+        expectedFiles = []
+        for seconds in ['00010','00020']:
+            expectedFiles.append('{}/restarts/restart.0001-01-01_{}.nc'
+                                 .format(self.sf.absdir, seconds))
+        self.assertEqual(files, expectedFiles)
+
+        files = self.sf.readpath('mesh', 'filename_template')
+        expectedFiles = ['mesh.nc']
+
```
Comment (pwolfram) [inline mpas_analysis/test/test_namelist_streams_interface.py]: @xylar, is an `self.assertEqual(files, expectedFiles)` missing?

File: mpas_analysis/test/test_namelist_streams_interface.py [hunk 8]
```diff
@@ -47,4 +49,54 @@ def test_read_streamsfile(self):
...
+
+        files = self.sf.readpath('mesh', 'filename_template',
+                                    startDate='0001-01-01',
+                                    endDate='0001-12-31')
+        expectedFiles = ['mesh.nc']
+
```
Comment (pwolfram) [inline mpas_analysis/test/test_namelist_streams_interface.py]: @xylar, is an `self.assertEqual(files, expectedFiles)` missing?

File: mpas_analysis/ocean/ocean_modelvsobs.py (hunk 2 above)
Comment (pwolfram) [inline mpas_analysis/ocean/ocean_modelvsobs.py]: Thanks @xylar, this is great.  I'm liking what I'm seeing so far.  The only substantive comment following this discussion is related to pre-filtering timesteps via custom parsing when xarray should be able to do this natively.  I don't think this is needed yet (or ever) but if I'm missing something please correct me.

**Comment by pwolfram** [comment]
@xylar, please let me know when you want a formal review, but so far things are looking great!

File: mpas_analysis/configuration/MpasAnalysisConfigParser.py [hunk 9]
```diff
@@ -26,7 +26,7 @@ def getWithDefault(self, section, option, default, listType=None):
         if isinstance(default, list):
             self.setlist(section, option, default)
         else:
-            return self.get(section, option, str(default))
+            self.set(section, option, str(default))
```
Comment (xylar) [inline mpas_analysis/configuration/MpasAnalysisConfigParser.py]: Yes, I agree this should be a separate PR.

File: mpas_analysis/test/test_namelist_streams_interface.py (hunk 7 above)
Comment (xylar) [inline mpas_analysis/test/test_namelist_streams_interface.py]: @pwolfram, yep, you're right.  Just noticed that myself.

**Comment by milenaveneziani** [comment]
@xylar: sorry you had to go through the trouble of sorting out the nco thing. That is the problem with being 8 hours ahead of time wrt us.. :) 
Yes, we need to use the newer version of nco, because that has the ncremap tool. It isn't a problem in ACME because we load the right nco module within the run*.csh script.
On edison, I set the nco/bin and library path in my bashrc.ext file: you can look at it, if it is useful.

File: mpas_analysis/ocean/sst_timeseries.py [hunk 10]
```diff
@@ -1,23 +1,47 @@
...
 
     casename = config.get('case','casename')
     ref_casename_v0 = config.get('case','ref_casename_v0')
     indir_v0data = config.get('paths','ref_archive_v0_ocndir')
 
-    compare_with_obs = config.getboolean('sst_timeseries','compare_with_obs')
+    #compare_with_obs = config.getboolean('sst_timeseries','compare_with_obs')
```
Comment (milenaveneziani) [inline mpas_analysis/ocean/sst_timeseries.py]: could we please leave this in for the moment? I am not sure whether we may need it (we have it in the seaice_modelvsobs script for sure).

Comment (xylar) [inline mpas_analysis/ocean/sst_timeseries.py]: @milenaveneziani, it's never used in this script, which is why I commented it out.  The script runs fine without it and produces bit-identical images.

Comment (milenaveneziani) [inline mpas_analysis/ocean/sst_timeseries.py]: you are right, we are not using that flag anymore. I should go back and do some clean-up at some point (we still have compare_with_model and compare_with_obs in the config file). no biggie for now.

**Comment by milenaveneziani** [comment]
@xylar: do you want me to test this on edison again?

**Comment by xylar** [comment]
@milenaveneziani, not quite yet.  I was writing on my laptop so I need to test it myself on Edison first.

**Comment by xylar** [comment]
@milenaveneziani and @pwolfram, okay this seems to work again.  I tested alpha7 and got bit-for-bit identical plots to `develop` with the config file at:
https://gist.github.com/xylar/228bdcd9ef4d19fd823ec33525e1ec42
I am still testing alpha8.  I got some errors about an empty array when I tried setting `timeseries_yr1 = 6` and `timeseries_yr2 = 10` but it seems to work as expected (so far) with the full range of data.

Anyway, feel free to test this yourselves and make further suggestions.  There may still be some bugs to work out, so it would be important to exercise this now, and continue doing so once we merge to `develop` before we do any harm to the ACME analysis.

Off to bed...

**Comment by xylar** [comment]
I tested alpha8 successfully.  The results (including the config file) are here:
http://portal.nersc.gov/project/acme/xylar/test_alpha8/

When I get a chance, I'll take a look at why other choices for `timeseries_yr1` and `timeseries_yr2` didn't work.  But I would say this is ready to merge into `develop` because the standard functionality where we look at the full range of data in time series analysis is working as expected.

@milenaveneziani, could you merge this as soon as #41 has been fulfilled and as soon as @pwolfram says he's okay with the code, tests, etc.?  Also, please make sure `pytest` passes for you as it did for me.

**Comment by pwolfram** [comment]
@xylar, so as long as we don't break the current workflow this should be merged to develop with an issue made for the bug you outline above, correct?

**Comment by xylar** [comment]
@pwolfram, that's correct.  In order to make a useful issue, I need to investigate a bit further.  I'll try to find time to do that tomorrow.

**Comment by pwolfram** [comment]
Tested using `module load nco/4.6.0`, but ran into a permissions issue @xylar:

```bash
┌─[pwolfram][edison12][~/xylarTest][13:04][±][read_paths_from_streams_file ✗]
└─▪ python run_analysis.py config.alpha8 
Traceback (most recent call last):
  File "run_analysis.py", line 157, in <module>
    analysis(config)
  File "run_analysis.py", line 68, in analysis
    makedirs(config.get('paths','plots_dir'))
  File "run_analysis.py", line 31, in makedirs
    os.makedirs(inpath)
  File "/global/project/projectdirs/acme/software/anaconda_envs/edison/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/global/project/projectdirs/acme/software/anaconda_envs/edison/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/global/project/projectdirs/acme/software/anaconda_envs/edison/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/scratch1/scratchdirs/xylar/acme'
```

Can you please fix the permissions @xylar and I can verify this works?

File: mpas_analysis/test/test_date.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        # YYYY-MM-DD_hh:mm:ss
+        # YYYY-MM-DD_hh.mm.ss
+        # YYYY-MM-DD_SSSSS
+        # DDD_hh:mm:ss
+        # DDD_hh.mm.ss
+        # DDD_SSSSS
+        # hh.mm.ss
+        # hh:mm:ss
+        # YYYY-MM-DD
+        # SSSSS
```
Comment (pwolfram) [inline mpas_analysis/test/test_date.py]: Nice @xylar

File: mpas_analysis/shared/io/namelist_streams_interface.py [hunk 12]
```diff
@@ -82,20 +98,112 @@ class StreamsFile:
...
+        on or before the endDate are included in the file list.
+        """
+        template = self.read(streamName, 'filename_template')
+        replacements = {'$Y':'[0-9][0-9][0-9][0-9]',
+                        '$M':'[0-9][0-9]',
+                        '$D':'[0-9][0-9]',
+                        '$S':'[0-9][0-9][0-9][0-9][0-9]',
+                        '$h':'[0-9][0-9]',
+                        '$m':'[0-9][0-9]',
+                        '$s':'[0-9][0-9]'}
```
Comment (pwolfram) [inline mpas_analysis/shared/io/namelist_streams_interface.py]: Nice @xylar

File: mpas_analysis/ocean/ocean_modelvsobs.py [hunk 13]
```diff
@@ -14,23 +14,38 @@
...
+
+    streams_filename = config.get('input', 'ocean_streams_filename')
+    streams = StreamsFile(streams_filename, streamsdir=indir)
+
+    # get a list of timeSeriesStats output files from the streams file,
+    # reading only those that are between the start and end dates
+    startDate = config.get('time', 'climo_start_date')
+    endDate = config.get('time', 'climo_end_date')
+    infiles = streams.readpath('timeSeriesStatsOutput',
+                               startDate=startDate, endDate=endDate)
```
Comment (pwolfram) [inline mpas_analysis/ocean/ocean_modelvsobs.py]: Why not inline this so that

```python
infiles = streams.readpath('timeSeriesStatsOutput',
                                          startDate=config.get('time', 'climo_start_date'), 
                                          endDate=config.get('time', 'climo_end_date'))
```

This may be cleaner, but that is arguable.  Note, I couldn't align the example above in markdown.

File: run_analysis.py [hunk 14]
```diff
@@ -88,7 +112,7 @@ def analysis(config): #{{{
         print "Plotting 2-d maps of SST climatologies..."
         from mpas_analysis.ocean.ocean_modelvsobs import ocn_modelvsobs
         ocn_modelvsobs(config, 'sst')
-
```
Comment (pwolfram) [inline run_analysis.py]: Thanks @xylar for this cleanup!

**Comment by pwolfram** [comment]
@xylar, please ignore the permissions query above.  The error was on my end and after talking with @milenaveneziani testing is in progress.

**Comment by milenaveneziani** [comment]
I am testing this on my side also, using first the alpha7 case. Unfortunately for this case the namelist and streams files are not in the same directory as the output files (because the original alpha7 is made up of two different restart simulations, and the current output directory was created for conveniency). I do not have permission to cp the namelists/streams in the new location.
How did you solve this problem with alpha7 @xylar?

**Comment by xylar** [comment]
> I am testing this on my side also, using first the alpha7 case. Unfortunately for this case the namelist and streams files are not in the same directory as the output files (because the original alpha7 is made up of two different restart simulations, and the current output directory was created for conveniency). I do not have permission to cp the namelists/streams in the new location.
How did you solve this problem with alpha7 @xylar?

@milenaveneziani, if you give an absolute path to the namelist and streams files (see the example `config.alpha7` at http://portal.nersc.gov/project/acme/xylar/test_alpha7/), they don't have to be in the same path as the files themselves.  That was a pain in the butt to add to the code, so I hope this isn't the only time it's useful...

**Comment by milenaveneziani** [comment]
ah, that's a wonderful idea @xylar. I am sure there would be plenty of other circumstances when the code would break without this. Thanks!

File: mpas_analysis/ocean/ocean_modelvsobs.py (hunk 13 above)
Comment (xylar) [inline mpas_analysis/ocean/ocean_modelvsobs.py]: Dammit!  I didn't want to do a review but I accidentally clicked to start one.  Then, when I cancelled my review it deleted my comment...

Here's the best re-creation I can come up with:
I'd rather leave it the way it is for two reasons:
* I find ii more readable my way
* it's possible we will want to reuse the `startDate` and `endDate` variables (e.g. to set more precise bounds on xarray in the future) and I like avoiding multiple `config.get` operations for the same variable whenever possible.

**Comment by xylar** [comment]
@milenaveneziani, good that we ran into this now, because otherwise I wouldn't have bothered to support this case ;-)

**Comment by xylar** [comment]
@pwolfram, as you probably noticed, I rebased on to `develop` to get the CI testing.  Worked like a charm!

**Comment by pwolfram** [comment]
@xylar and @milenaveneziani, I tested this with alpha7 and alpha8 and it worked.  LGTM

**Comment by milenaveneziani** [comment]
alpha7 test passed with timeseries=1-9999. Testing alpha8 now.
I will merge this after that and then we will look in the timeseries_yr# problem later?

**Comment by milenaveneziani** [comment]
All tests passed multiple times for both alpha7 and alpha8.
Thanks @xylar: this was a great addition.
